orderer/common/multichannel/registrar.go

```diff
@@ -463,18 +535,37 @@ func (r *Registrar) JoinChannel(channelID string, configBlock *cb.Block, isAppCh
 		return types.ChannelInfo{}, errors.Wrap(err, "failed extracting config envelope from block")
 	}
 
-	//TODO save the join-block in the file repo to make this action crash tolerant.
+	blockBytes, err := proto.Marshal(configBlock)
+	if err != nil {
+		return types.ChannelInfo{}, errors.Wrap(err, "failed marshaling joinblock")
+	}
+
+	if err := r.joinBlockFileRepo.Save(channelID, blockBytes); err != nil {
+		return types.ChannelInfo{}, errors.Wrapf(err, "failed saving joinblock to file repo for channel %s", channelID)
+	}
 
 	ledgerResources, err := r.newLedgerResources(configEnv)
 	if err != nil {
-		//TODO remove join block
-		return types.ChannelInfo{}, errors.Wrap(err, "failed creating ledger resources")
+		err = errors.Wrap(err, "failed creating ledger resources")
+
+		if err2 := r.removeJoinBlock(channelID); err2 != nil {
+			// Stack file repo error on top of ledger resource creation error
+			err = errors.WithStack(err2)
+		}
+
+		return types.ChannelInfo{}, err
 	}
 
 	joinSupport, err := newChainSupportForJoin(configBlock, r, ledgerResources, r.consenters, r.signer, r.blockcutterMetrics, r.bccsp)
 	if err != nil {
-		//TODO remove join block, clean ledger resources
-		return types.ChannelInfo{}, errors.Wrap(err, "failed creating chain support for join")
+		err = errors.Wrap(err, "failed creating chain support for join")
+
+		if err2 := r.removeJoinBlock(channelID); err2 != nil {
+			// Stack file repo error on top of chain support creation error
+			err = errors.WithStack(err2)
+		}
+
+		return types.ChannelInfo{}, err
```

Comment:
at this point we have to clean the ledger as well

Reply:
Done